iravelmakina/caesar-encryption-algorithm-ivelmakina
# Caesar Encryption Algorithm

## Description
- Added Caesar algorithm dynamic library
- Added Caesar algorithm static library
- Implemented dynamic library testing
- Built a DLL and executable for dynamic library
- Built a LIB and executable for static library
- Built a program utilizing `caesar.dll`

## Changes
- Created `caesar.cpp` and `main.cpp` for future compilation and linking into DLL
  - `caesar.cpp` includes `encrypt()` and `decrypt()` functions
  - `main.cpp` includes calls to functions in `caesar.cpp`
- Created `caesar.cpp` and `main.cpp` for future compilation and linking into LIB
  - `caesar.cpp` includes `encrypt()` and `decrypt()` functions
  - `main.cpp` includes calls to functions in `caesar.cpp`
- Created a simple CLI program exporting the functions from `caesar.dll` to encrypt and decrypt messages using keys
  - Added helper function `toLowerCase()`
  - Added functions `getAndValidateUserInput()` and `processMessage()`
  - Added function `loadLibrary()`
  - Added functions `runProgram()` and `main()`
- Compiled and linked `caesar.cpp` (`caesar.o` -> `caesar.dll`) and `main.cpp` (`main.o` -> `main.exe`)
- Compiled and linked `caesar.cpp` (`caesar.o` -> `caesar.lib`) and `main.cpp` (`main.o` -> `main.exe`)
- Compiled and linked `program_cli.cpp` (`program_cli.o` -> `program_cli.exe`)

## How to Test
1. Compile `caesar.cpp` and `main.cpp` to create `caesar.dll` and `main.exe`
2. Compile `caesar.cpp` and `main.cpp` to create `caesar.lib` and `main.exe`
3. Compile `program_cli.cpp` to create `program_cli.exe`
4. Run `program_cli.exe` and test the encryption and decryption functions using different keys and messages
5. Validate the output to ensure it matches the expected results for both encryption and decryption

## Screenshots
1.
![dll1](https://github.com/iravelmakina/caesar-encryption-algorithm-ivelmakina/assets/150923968/89491cad-5462-44bb-8c0a-b3595b9341de)
![dll2](https://github.com/iravelmakina/caesar-encryption-algorithm-ivelmakina/assets/150923968/c72a7e4c-5e84-4e1a-8bc8-e5ad68085132)
![dll3](https://github.com/iravelmakina/caesar-encryption-algorithm-ivelmakina/assets/150923968/98373010-56e8-4cda-bf5f-d1b1bf8be56e)
![dll4](https://github.com/iravelmakina/caesar-encryption-algorithm-ivelmakina/assets/150923968/e60548b2-fb03-42c6-8504-09e19d45ccf2)
![dll5](https://github.com/iravelmakina/caesar-encryption-algorithm-ivelmakina/assets/150923968/1054e2f0-4619-4a20-8262-c995eb7cfe1f)
![dll6](https://github.com/iravelmakina/caesar-encryption-algorithm-ivelmakina/assets/150923968/88a66647-26e3-4e20-b528-183f38faef69)
![image](https://github.com/iravelmakina/caesar-encryption-algorithm-ivelmakina/assets/150923968/7f160f0a-d011-4d4a-99e9-f6b9b0d27e88)

2.
![lib1](https://github.com/iravelmakina/caesar-encryption-algorithm-ivelmakina/assets/150923968/05d1353f-58c7-4393-acac-f2b5702b1e8b)
![lib2](https://github.com/iravelmakina/caesar-encryption-algorithm-ivelmakina/assets/150923968/c977c068-c158-4283-bef8-434366202498)
![lib3](https://github.com/iravelmakina/caesar-encryption-algorithm-ivelmakina/assets/150923968/c2e89915-353d-4ce9-bec7-453a779eda08)
![lib4](https://github.com/iravelmakina/caesar-encryption-algorithm-ivelmakina/assets/150923968/2741c049-e854-480a-962a-f49c4f4631c9)
![image](https://github.com/iravelmakina/caesar-encryption-algorithm-ivelmakina/assets/150923968/22620abc-98fc-4ebb-8c8e-9eee133988f9)


3-5.
![program1](https://github.com/iravelmakina/caesar-encryption-algorithm-ivelmakina/assets/150923968/df964cfc-16f5-42c6-bf74-f514298079ab)
![program2](https://github.com/iravelmakina/caesar-encryption-algorithm-ivelmakina/assets/150923968/54d6cfd6-cb5f-43cb-af9d-87b4a5e59b2c)
![program3](https://github.com/iravelmakina/caesar-encryption-algorithm-ivelmakina/assets/150923968/d113e70f-e309-418f-af18-a1625d61e4be)
![program4](https://github.com/iravelmakina/caesar-encryption-algorithm-ivelmakina/assets/150923968/322e8462-3ef7-475c-a33f-df48092e69cc)
![image](https://github.com/iravelmakina/caesar-encryption-algorithm-ivelmakina/assets/150923968/564727e5-6611-46d9-bdba-825a5c3546ce)
